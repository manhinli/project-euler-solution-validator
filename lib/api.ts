import { NextApiResponse } from "next";

type Json = Record<string, unknown> | Json[];

/**
 * Sends HTTP 200 OK back to client with given data.
 *
 * @param res Next.js response object
 * @param data Data to send to client
 */
export function respondSuccessWithJson(res: NextApiResponse, data: Json) {
    return res.status(200).json(data);
}

/**
 * Sends a client error response (defaulting to HTTP 400) with given error.
 *
 * @param res Next.js response object
 * @param code Response code
 * @param error Error object
 */
export function respondClientError(
    res: NextApiResponse,
    code: number = 400,
    error?: unknown
) {
    let message = "Bad request";

    // Determine if `error` has a message that we can provide
    if (error && typeof error === "object" && Object.hasOwn(error, "message")) {
        message = `${(error as any).message}`;
    } else if (typeof error === "string") {
        message = error;
    }

    return res.status(code).json({
        error: message,
    });
}

/**
 * Sends HTTP 404 Not Found back to client.
 *
 * @param res Next.js response object
 */
export function respondNotFound(res: NextApiResponse) {
    return respondClientError(res, 404, "Not found");
}

/**
 * Sends HTTP 405 Method Not Allowed back to client.
 *
 * @param res Next.js response object
 */
export function respondMethodNotAllowed(res: NextApiResponse) {
    return respondClientError(res, 405, "Method not allowed");
}
