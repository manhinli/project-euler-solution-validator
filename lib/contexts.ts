import React, { Dispatch, SetStateAction } from "react";

export const UsernameContext = React.createContext<{
    value: string;
    onChange: Dispatch<SetStateAction<string>>;
}>({
    value: "",
    onChange: () => {},
});
