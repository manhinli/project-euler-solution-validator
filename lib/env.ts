export const ENVIRONMENT_VARIABLES = [
    "DATABASE_URI",
    "DATABASE_NAME",
    "PROBLEMS_COMPILED_PATH",
] as const;

export function getEnv(variable: typeof ENVIRONMENT_VARIABLES[number]) {
    const value = process.env[variable];

    if (!value) {
        throw new Error(`${variable} not defined`);
    }

    return value;
}
