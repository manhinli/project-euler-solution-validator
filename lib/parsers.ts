import { z } from "zod";

/**
 * Parses a value that is intended to be a problem ID.
 *
 * @param id Value to be parsed as problem ID
 * @returns {number} Problem ID value
 */
export function parseProblemId(id: any): number {
    const schema = z
        .number()
        .int("Problem ID must be integer")
        .positive("Problem ID must be positive integer");

    return (
        z
            // Apply value transforms before testing against schema
            .preprocess((value) => {
                // If already number, continue
                if (typeof value === "number") {
                    return value;
                }

                // If string, parse as number
                if (typeof value === "string") {
                    return Number.parseFloat(value);
                }

                // Otherwise ignore value
                return undefined;
            }, schema)
            // Parse value
            .parse(id)
    );
}
