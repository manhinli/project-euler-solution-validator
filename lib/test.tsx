import "@testing-library/jest-dom";
import { SWRConfig } from "swr";

// Provide a fresh provider every time we use SWR
export const swrConfigWrapper = (
    children: any,
    config: React.ComponentProps<typeof SWRConfig>["value"] = {}
) => (
    <SWRConfig value={{ provider: () => new Map(), ...config }}>
        {children}
    </SWRConfig>
);

// Utility to deep clone objects
export const deepClone = (o: object) => JSON.parse(JSON.stringify(o));
