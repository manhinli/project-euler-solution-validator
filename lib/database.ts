import { Collection, MongoClient } from "mongodb";
import { getEnv } from "./env";

export const DATABASE_URI = getEnv("DATABASE_URI");
export const DATABASE_NAME = getEnv("DATABASE_NAME");
export const COLLECTIONS = ["attempts", "problems"] as const;

// Singleton MongoDB client instance
let mongoClient: MongoClient | undefined = undefined;

export async function getClient() {
    if (mongoClient) {
        return mongoClient;
    }

    const client = new MongoClient(DATABASE_URI);
    mongoClient = await client.connect();
    return mongoClient;
}

export async function getDatabase() {
    const client = await getClient();
    return client.db(DATABASE_NAME);
}

export async function getCollection<T>(
    collection: typeof COLLECTIONS[number]
): Promise<Collection<T>> {
    const db = await getDatabase();
    return db.collection(collection);
}
