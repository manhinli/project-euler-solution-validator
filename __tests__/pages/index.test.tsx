import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Index from "../../pages";
import { swrConfigWrapper } from "../../lib/test";
import { act } from "react-dom/test-utils";

describe("Index", () => {
    it("renders welcome", () => {
        render(<Index />);

        expect(screen.getByText("Welcome")).toBeInTheDocument();
    });

    it("renders problems fetched from server", async () => {
        const fetcher = jest.fn(() =>
            Promise.resolve([
                {
                    problemId: 42,
                    title: "Ultimate Question of Life, the Universe, and Everything",
                },
                {
                    problemId: 85,
                    title: "Number of Known Primes that existed before 1984",
                },
            ])
        );

        render(swrConfigWrapper(<Index />, { fetcher }));

        const table = screen.getByRole("table");

        // Wait for data to be processed and updated on the page
        await act(async () => {
            await Promise.resolve();
        });

        expect(table).toMatchSnapshot();
        expect(fetcher.mock.calls[0][0]).toBe("/api/problem");
    });
});
