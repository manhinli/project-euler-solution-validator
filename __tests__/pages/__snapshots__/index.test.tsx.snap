// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index renders problems fetched from server 1`] = `
<table
  class="problems"
>
  <thead>
    <tr>
      <th>
        ID
      </th>
      <th>
        Title
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        42
      </td>
      <td>
        <a
          href="/problem/42"
        >
          Ultimate Question of Life, the Universe, and Everything
        </a>
      </td>
    </tr>
    <tr>
      <td>
        85
      </td>
      <td>
        <a
          href="/problem/85"
        >
          Number of Known Primes that existed before 1984
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;
